src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -132,13 +133,15 @@ class Simplex_tree {
 
   struct Key_simplex_base_real {
     Key_simplex_base_real() : key_(-1) {}
+    Key_simplex_base_real(Simplex_key k) : key_(k) {}
     void assign_key(Simplex_key k) { key_ = k; }
     Simplex_key key() const { return key_; }
    private:
     Simplex_key key_;
   };
   struct Key_simplex_base_dummy {
     Key_simplex_base_dummy() {}
+    Key_simplex_base_dummy([[maybe_unused]] Simplex_key k) {}
```

Comment:
Actually, isn't it a bit dangerous to allow this and silently ignore the argument? `assign_key` exists but using it causes an error. For dummy filtrations we have an assertion that we don't assign a non-default value.

Reply:
I already found it weird that `Filtration_simplex_base_dummy` and `Key_simplex_base_dummy` work so differently. The filtration dummy allows to use `assign_filtration` and `filtration` in release mode and `assign_filtration` throws in debug mode if you try to assign something else than 0. Meanwhile, the key dummy throws systematically if you try to call `assign_key` or `key`. So I am a bit confused on what to do.

But I can indeed already add `GUDHI_CHECK` in both constructors.

By the way, `GUDHI_CHECK_code` does not seem to do much for now, no? Is it necessary?

Reply:
`assert(n>=0)` doesn't mean that `n<0` is allowed in release mode.
`GUDHI_CHECK_code` is for code that is only used by a later GUDHI_CHECK.
The current state is the result of a compromise. On the one hand, when the class has no filtration/key, we want the user to realize quickly that something is wrong is they try to use it. On the other hand, we don't want to write `if constexpr` every other line (and the code predates C++17, so it was even worse), so it can be convenient to have dummy functions. To reconcile the 2 points, the dummy functions produce an error if used, preferably at compile time, otherwise at runtime (in debug mode to avoid slowing down release mode).
Also, filtration and key have different semantics: with no key, we have no key, that's it, while with no filtration, we chose to interpret it as everything having the same filtration (arbitrarily: 0). Of course we could have made a different choice...

Reply:
> `assert(n>=0)` doesn't mean that `n<0` is allowed in release mode.

I know, it depends on the compiling flags if I remember well. But when compiling Gudhi in release mode, it didn't triggered the throw when I called `assign_filtration(n)` at some point during my tests, it only did it when I recompiled in Debug mode.

> On the other hand, we don't want to write `if constexpr` every other line (and the code predates C++17, so it was even worse),

But precisely because there is this difference, I had to use the `if constexpr (Options::store_key)` at different places. It would make more sense to do the same with the dummy key than with the dummy filtration values.

Reply:
Compiler flags and release/debug mode are the same thing. This is a question of perspective, but let me explain how I see things.
When we have `GUDHI_CHECK(f==0)` in a function, it means that the caller should **never** pass anything but `0` for `f`, whatever the compiler options. If you pass 1, in release mode, it may let you (it won't waste time checking), but whatever follows is undefined behavior; in debug mode, it may stop you, the way it stops you happens to be an exception, but I was against documenting it because this isn't like `vector::at()` where you can reliably catch the exception, here it is an early error that helps you debug your code by telling you about your mistake.

> It would make more sense to do the same with the dummy key than with the dummy filtration values.

Maybe. A compile-time error is preferable to a runtime one whenever possible, but if the runtime one really simplifies things (I didn't check)...

Reply:
> Compiler flags and release/debug mode are the same thing. 

Hm, I am not sure. If I remember well, you can use de release mode (which is a flag I know) and still have the assert working by using an additional flag.

> This is a question of perspective, but let me explain how I see things. When we have `GUDHI_CHECK(f==0)` in a function, it means that the caller should **never** pass anything but `0` for `f`, whatever the compiler options. If you pass 1, in release mode, it may let you (it won't waste time checking), but whatever follows is undefined behavior; in debug mode, it may stop you, the way it stops you happens to be an exception, but I was against documenting it because this isn't like `vector::at()` where you can reliably catch the exception, here it is an early error that helps you debug your code by telling you about your mistake.

I agree, but this changes nothing to my comment, no? For dummy filtration values, you **should not** use `assign_filtration`, whereas for dummy keys you **cannot** use `assign_key`. In the end, there is the possibility to handle them differently which is not completely coherent in my point of view (even if I get that they don't have the same role). And now, the dummy key is even less coherent as you can "assign" a value by using the constructor (even though not wanted as they is a GUDHI_CHECK), but assigning a value via `assign_key` always throws.

> 
> > It would make more sense to do the same with the dummy key than with the dummy filtration values.
> 
> Maybe. A compile-time error is preferable to a runtime one whenever possible, but if the runtime one really simplifies things (I didn't check)...

I also prefer a compile time error over a run time error. And I personally like the `if constexpr` even if they not always make things very readable. I would have preferred this solution even for the filtration values. It is just that you said that you don't like to use them when not necessary, and I can understand why. I was just highlighting that we use two opposite solutions to more or less the same problem...

Reply:
* "Release mode" is a more or less arbitrary set of flags (typically "-O3 -DNDEBUG" with gcc, it mixes "optimize", "no assertions", "no debug symbols"). Different flags (say just "-O3") can enable optimizations without disabling assertions, if that's what you mean.
* "assigning a value via assign_key always throws." please don't use "throws" (which seems to imply an exception thrown at runtime from this function) for "fails to compile/link" (what currently happens with the dummy assign_key). You don't really need `Key_simplex_base_dummy(Simplex_key)`, you can split out the constructor of Simplex_tree_node_explicit_storage that takes a key, this avoids making it "even less coherent".
* You favor more consistency, we went with something slightly inconsistent to get an earlier error for keys (where is was possible) / less verbose code for filtrations. Both approaches make sense.

Reply:
> * "Release mode" is a more or less arbitrary set of flags (typically "-O3 -DNDEBUG" with gcc, it mixes "optimize", "no assertions", "no debug symbols"). Different flags (say just "-O3") can enable optimizations without disabling assertions, if that's what you mean.

Yes, that is what I mean.

>     * "assigning a value via assign_key always throws." please don't use "throws" (which seems to imply an exception thrown at runtime from this function) for "fails to compile/link" (what currently happens with the dummy assign_key). You don't really need `Key_simplex_base_dummy(Simplex_key)`, you can split out the constructor of Simplex_tree_node_explicit_storage that takes a key, this avoids making it "even less coherent".

Right, "throw" is not the right word. It was just meant to make a difference between "should not" and "cannot".
And yes, I know that the constructor is not needed, I didn't used it first. But it felt like pushing the problem somewhere else and in case we wanted to change the current approach, it would have made more sense with the constructor. But I know that it is subjective, I have no problem to change it back.

>     * You favor more consistency, we went with something slightly inconsistent to get an earlier error for keys (where is was possible) / less verbose code for filtrations. Both approaches make sense.

There is perhaps a misunderstanding here. I made my initial remark only because I was not sure how I was suppose to handle the changes in those two classes as they work so differently while being so similar. Just to make sure: there was no criticism here. I wanted to make sure that the current behaviors were wanted and not just artifacts which went under the radar. So, if I understand well: I should continue to push key errors as much as possible to compile time while keeping filtration value errors at run time? Even if I find it weird, I am fine going with it.

Reply:
> I wanted to make sure that the current behaviors were wanted and not just artifacts which went under the radar.

Maybe not strongly wanted, but they were conscious choices, yes.

> So, if I understand well: I should continue to push key errors as much as possible to compile time while keeping filtration value errors at run time?

For the time being, I would say yes, unless you can show a significant advantage to changing it (maybe such an advantage will appear naturally after changing the meaning of insert, or some other future change, I don't know).